### PR TITLE
Add path properties endpoint (TS2)

### LINKED ISSUE #6785: TS2: Add editoast path properties endpoint
## Description

Add an endpoint `POST` `/v2/infra/ID/path_properties?props[]=slopes&props[]=gradients&props[]=electrification&props[]=neutral_sections&props[]=geometry`

We should be able to query these properties:
- slopes
- gradients
- electrification & neutral_sections (we should check if we depend on the rolling stock)
- geometry
- operational_points

## AC

- [ ] The endpoint, should cache the result. (We shouldn't do more than one redis call) 
- [ ] We should only return the queried properties
- [ ] The computation should be done by `core` (no need to implement core endpoint yet)
- [ ] Use `serde_qs` to deserialize queyparams element

### PR DESCRIPTION
**Pull request by younesschrifi**
Close #6785 

Multiple smoke tests done to ensure the cache behaves as expected (but not reproduced in unit tests)

### CODE REVIEW

**Review by flomonster — APPROVED** [review]

File: editoast/src/schema/utils/duration.rs [hunk 1]
```diff
@@ -117,13 +117,13 @@ where
...
-                + ChronoDuration::seconds(second as i64)
-                + ChronoDuration::milliseconds(millisecond as i64)
+            ChronoDuration::try_days(day as i64).unwrap()
+                + ChronoDuration::try_hours(hour as i64).unwrap()
+                + ChronoDuration::try_minutes(minute as i64).unwrap()
+                + ChronoDuration::try_seconds(second as i64).unwrap()
+                + ChronoDuration::try_milliseconds(millisecond as i64).unwrap()
         }
-        IsoDuration::Weeks(weeks) => ChronoDuration::weeks(weeks as i64),
+        IsoDuration::Weeks(weeks) => ChronoDuration::try_weeks(weeks as i64).unwrap(),
```
Comment (woshilapin) [inline editoast/src/schema/utils/duration.rs]: Why unwrapping in a function that already returns `Result`? It might be as easy as replacing the `.unwrap()` with `.ok_or_else(|| serde::de::custom("value for millisecond is not valid"))`.

Aside note: I'm really surprised by `chrono`'s new API to have functions prefixed with `try_` that do not return a `Result` 🤯 Seems like a really weird/uncommon choice.

File: editoast/src/views/v2/path.rs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+#[editoast_error(base_id = "pathfinding")]
+enum PathfindingError {
+    #[error("Infra '{infra_id}', could not be found")]
+    #[editoast_error(status = 404)]
+    InfraNotFound { infra_id: i64 },
+}
+
+async fn retrieve_infra_version(conn: &mut PgConnection, infra_id: i64) -> Result<String> {
+    use crate::models::Infra;
+    use crate::models::Retrieve;
```
Comment (woshilapin) [inline editoast/src/views/v2/path.rs]: Should we put these import at the beginning of the file? It's rather uncommon to import types/traits inside a scope. Another solution, if they're used only once in the entire file is to inline it directly (might be less ideal for the train though 🤔).

EDIT: I found out later in my review that it's actually code that only has been moved from another file. But it might be the opportunity to correct it if we want to.

File: editoast/src/redis_utils.rs [hunk 3]
```diff
@@ -72,9 +72,9 @@ impl RedisConnection {
     pub async fn json_get_ex<T: DeserializeOwned, K: ToRedisArgs + Send + Sync>(
         &mut self,
         key: K,
-        seconds: usize,
+        seconds: u64,
```
Comment (woshilapin) [inline editoast/src/redis_utils.rs:75]: I don't understand why this change was needed. The function itself need a `usize`, so it would be more logical that the caller of this function does the conversion.

**Review by Wadjetz — APPROVED** [review]

File: editoast/src/views/v2/path/properties.rs [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+            properties.insert(Property::Geometry);
+        }
+        if self.operational_points.is_some() {
+            properties.insert(Property::OperationalPoints);
+        }
+        properties
+    }
+
+    /// Merge another path properties into this one
+    fn merge(self, other: PathProperties) -> Self {
```
Comment (woshilapin) [inline editoast/src/views/v2/path/properties.rs:85]: From the name of the function and the types involved (mostly fields with `Option`), I would expect `self` to be kept in priority and `other` to be used only if needed. But it's actually the opposite that happens. Rather misleading. See my comment below, both might be complementary to find a coherent way of writing all of this.

File: editoast/src/views/v2/path/properties.rs [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+            slopes: other.slopes.or(self.slopes),
+            gradients: other.gradients.or(self.gradients),
+            electrifications: other.electrifications.or(self.electrifications),
+            geometry: other.geometry.or(self.geometry),
+            operational_points: other.operational_points.or(self.operational_points),
+        }
+    }
+
+    /// Filter properties not requested
+    pub fn filter_properties(&mut self, properties: Properties) {
```
Comment (woshilapin) [inline editoast/src/views/v2/path/properties.rs]: ⚠️ I'll make a very subjective opinion here so take with a grain of salt.

I'd say using `&mut self` functions should be reduced to the minimum. Functions with side effects are often more difficult to reason about in a large or complex piece of code. Sometimes, you don't have a choice, but here, maybe approaching this with a more functional paradigm would be beneficial.

For example, the function could be written this way.

```rust
pub fn filter_properties(mut self, properties: Properties) -> Self { /* ... */ }
```

Then, when you use it, you could chain the calls a bit more.

```rust
let path_properties = computed_path_properties.filter_properties(missing_props).merge(path_properties);
```

Note that the behavior of `merge` might need to be changed, but see my comment on the `merge` function, it might actually be a good thing to invert its behavior.

File: editoast/src/views/v2/path/properties.rs [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+#[serde(rename_all = "snake_case")]
+enum Property {
+    Slopes,
+    Gradients,
+    Electrifications,
+    Geometry,
+    OperationalPoints,
+}
+
+/// Alias `EnumSet<Property>`
```
Comment (woshilapin) [inline editoast/src/views/v2/path/properties.rs]: The comment is not super useful. Can be removed.

File: editoast/src/views/v2/path/properties.rs [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+    if let Some(electrifications) = &path_properties.electrifications {
+        redis_conn
+            .json_set_ex(&elec_hash, electrifications, CACHE_PATH_EXPIRATION)
+            .await?;
+    };
+
+    Ok(())
+}
+
+/// Compute path properties input hash without suported electrifications
```
Comment (woshilapin) [inline editoast/src/views/v2/path/properties.rs]:
```suggestion
/// Compute path properties input hash without supported electrifications
```

**Review by woshilapin — APPROVED** [review]
Sorry for some very long comment, I preferred to explain better what I had in mind instead of being too concise.

Overall, the PR is good and most of my comments are small remarks, nothing blocking from merging as-is.

File: editoast/src/redis_utils.rs (hunk 3 above)
Comment (flomonster) [reply editoast/src/redis_utils.rs:75]: Redis has an inconsistent API. It uses `u64` for `set_ex` but `usize` for `get_ex`. I wanted to unify this interface.

File: editoast/src/schema/utils/duration.rs [hunk 8]
```diff
@@ -117,13 +117,19 @@ where
...
             }
-            ChronoDuration::days(day as i64)
-                + ChronoDuration::hours(hour as i64)
-                + ChronoDuration::minutes(minute as i64)
-                + ChronoDuration::seconds(second as i64)
-                + ChronoDuration::milliseconds(millisecond as i64)
+            ChronoDuration::try_days(day as i64)
+                .ok_or_else(|| serde::de::Error::custom("value for millisecond is not valid"))?
+                + ChronoDuration::try_hours(hour as i64)
+                    .ok_or_else(|| serde::de::Error::custom("value for millisecond is not valid"))?
```
Comment (flomonster) [inline editoast/src/schema/utils/duration.rs]: The error message should be updated.

```suggestion
                    .ok_or_else(|| serde::de::Error::custom("value for hours is not valid"))?
```

**Review by flomonster — APPROVED** [review]

File: editoast/src/schema/utils/duration.rs [hunk 9]
```diff
@@ -117,13 +117,19 @@ where
...
+                + ChronoDuration::try_minutes(minute as i64)
+                    .ok_or_else(|| serde::de::Error::custom("value for minutes is not valid"))?
+                + ChronoDuration::try_seconds(second as i64)
+                    .ok_or_else(|| serde::de::Error::custom("value for seconds is not valid"))?
+                + ChronoDuration::try_milliseconds(millisecond as i64)
+                    .ok_or_else(|| serde::de::Error::custom("value for millisecond is not valid"))?
         }
-        IsoDuration::Weeks(weeks) => ChronoDuration::weeks(weeks as i64),
+        IsoDuration::Weeks(weeks) => ChronoDuration::try_weeks(weeks as i64)
+            .ok_or_else(|| serde::de::Error::custom("value for millisecond is not valid"))?,
```
Comment (woshilapin) [inline editoast/src/schema/utils/duration.rs:132]: ooops 😉 The next one doing a PR on `editoast` can take that one 😄 

```suggestion
            .ok_or_else(|| serde::de::Error::custom("value for week is not valid"))?,
```

Comment (woshilapin) [reply editoast/src/schema/utils/duration.rs:132]: Done in https://github.com/osrd-project/osrd/pull/6842. Thanks @younesschrifi
